src/main/java/com/google/devtools/build/lib/pkgcache/TestFilter.java

```diff
@@ -107,7 +109,7 @@ public boolean equals(Object o) {
     TestFilter f = (TestFilter) o;
     return f.testSizeFilterSet.equals(testSizeFilterSet)
         && f.testTimeoutFilterSet.equals(testTimeoutFilterSet)
-        && f.testTagFilterList.equals(testTagFilterList)
+        && Objects.equals(f.testTagFilter, testTagFilter)
```

Comment:
Since Starlark expressions don't have defined equality, can you plumb the original String through instead and use that for comparison?

Then in this class, you can parse the String into a /transient/ Expression. Also you can remove the AutoCodec annotations.

To be able to surface expression errors early, you would need to parse twice but that shouldn't be an issue.

Reply:
Expression strings are not precisely defined, and they abbreviate the full syntax tree. (At a guess, this feature was added so that stack traces could print approximate source syntax of the call stack; that's no longer needed now that we read and print the actual source.) So Expression.toString isn't suitable for this purpose: two different expressions may have the same string.

I suggest using the original expression provided by the user, parsing it once for validation, and again when needed.

I'm sorry the hidden constraints imposed by Google-internal code have made this PR such an ordeal for you. Thanks for your patience.

Reply:
Aha, thanks for the clarification! Then I will change to use the original string instead.